### DISCUSSION TITLE
New Training Method : Masked Score Estimation

### DISCUSSION THREAD

**Discussion by cloneofsimo** [cloneofsimo/lora#96] [General]
Basically copy-paste from my twitter thread, but sharing here as well:
Consider yourself fine-tuning on the following two images. As a neural network, you have no idea if blue cloth is something intrinsic to <Enid>. But what if you modified the score estimation objective to only consider the region of interest : in our case, facial features?

![image](https://user-images.githubusercontent.com/35953539/209966476-a2931875-6fd7-4e21-a64a-9c59b78da521.png)

Explicitly, we can define the region of interest as some appropriate mask function given image x_t. In my experiments, I've used result from face recognition model, and scaled them to make them have equal importance.

$$
\begin{align}
L &= \mathbb{E}_{t~[1, \lambda T]}||(\epsilon_t - \epsilon_\theta (x_t , t, c))\cdot M(x_t)||^2 \\
M(x_t) &= \frac{S(x_t) + \delta}{s(x_t)} \\
\lambda& \in [0,1]
\end{align}
$$

One way to make sense of this simple trick, is to see it as "identification" (or projection) with regards to equivalence relation : If the region of interest is same, they are the same. Here is the ablation study, Left 4 pictures contain "school-uniformness" and "blueness" on their cloth, whereas right 4 pictures have non of them. seed from 0 to 3.

![image](https://user-images.githubusercontent.com/35953539/209966938-92e7cb21-e5f7-4693-8763-882e6fd55fff.png)

Here are results with random prompts from lexica, NOT CHERRYPICKED.

![image](https://user-images.githubusercontent.com/35953539/209967171-4299e18d-77d7-40d8-b501-26b6147a27d1.png)

Currently merged into #88 .

**Comment by brian6091** [comment]
That's a great idea. Would be nice to build this into a general augmentation pipeline, or even to allow user-defined masks. I wonder if we could apply it to finetune the inpainting model?

**Reply by cloneofsimo** [reply]
Yeah, "face recognition module" is currently attached to the dataset themselves, but I think we can simply make it to have their own pipeline. I think it will work with inpainting as well.

**Reply by torealise** [reply]
Is it already implemented? Is it possible to train LoRA with mask? For example I have a woman in red t-shirt and blue jeans and I want to train LoRA so it remember only red t-shirt (using mask)?

**Comment by brian6091** [comment]
Also wonder how much of this could be acheived with captioning. E.g., "TOKEN wearing blue and black striped jacket" to get the model to dissociate those features from the target token.

**Reply by torealise** [reply]
So is there an answer?

**Comment by GLivshits** [comment]
Can you please describe in details the second formula (for M)? What are S and delta, and how to get them?

**Reply by cloneofsimo** [reply]
You can check out the code if you are interested, it is just very general formulation. In my case, S is the mask function + gaussian blur , so 1 where there is a face and 0 otherwise.  s is the sum of the S, so it is basically normalizing term (sometimes head size is different for all images, so I put it there) delta is just constant to make sure that not all backgrounds are entirely ignored. If you read the code, it will make sense much faster.

**Reply by andupotorac** [reply]
So if one wants to do the reverse, train on everything else but the face, they would swap the 0 and the 1? This would be useful for example when training on a style and not wanting to train on specific faces used in that style.

**Comment by cloneofsimo** [comment]
Throwing bunch of results I've shared on Twitter just for reference. All used face segmentation

![KakaoTalk_20230109_194300933](https://user-images.githubusercontent.com/35953539/211848291-81e8ef37-0357-479c-9f61-b2cc5a4a928e.png)
![Linaqruf_anything-v3 0_2](https://user-images.githubusercontent.com/35953539/211848365-a918ed44-af64-4099-a25f-eb4c36e5dc95.png)
![3](https://user-images.githubusercontent.com/35953539/211848378-d4bbc5bf-7787-439c-b92d-004d7e4cfdf1.png)
![image](https://user-images.githubusercontent.com/35953539/211848630-f492ff38-b074-40f8-9dad-7523dadf5f9b.png)
![image](https://user-images.githubusercontent.com/35953539/211848785-96ff7073-a945-44de-a990-4232ddbaf1ae.png)

**Reply by runner22k** [reply]
did anyone observe the difference in Bottom three faces? actual photo(1) and the output (3) right image doesn't look same. It looks odd for some reason. Can LoRA achieve perfection? How much training is needed to reach dreambooth quality? Sorry I haven't used LoRA much.

**Comment by krahnikblis** [comment]
for a work project i messed around with [this](https://eqxvision.readthedocs.io/en/latest/getting_started/Self-Attention%20Maps/) (which is JAX but based on [this](https://github.com/facebookresearch/dino) which is pytorch if you are sticking with that library) to demonstrate how attention models apply to product photography in digital advertising (we found a correlation with click-to-view rates and ROI vs the layout and concentration of these attention maps)... 

i think it would probably be somewhat helpful here as well, both for visualizing what the attention layers are "looking at", as well as to auto-define them as masks (i bet you could combine the concept with CLIP, such that the self-attention layers "focus" on the image areas matching a text prompt)

**Reply by brian6091** [reply]
That's cool. Completely agree about CLIP, especially when I saw the they got [CLIPSeg in transformers now](https://huggingface.co/blog/clipseg-zero-shot). I've been playing with a kind of soft-prompting, where instead of captioning an image with just one, I draw randomly from several (up to five depending on how much trouble I'm having using a single caption to describe an image), and I think it would be cool to use CLIPSeg to automask the images

**Reply by cloneofsimo** [reply]
https://github.com/7eu7d7/DreamArtist-sd-webui-extension#attention-mask might be relevant!

**Reply by krahnikblis** [reply]
actually, as soon as you think of a thing, someone else already published a paper on it. this is a LOT of fun, especially to help understand what the attention heads/layers are doing, word by word. https://github.com/castorini/daam - check the colab they link, it has a neat per-head viz option too, in addition to per-word. i altered that a little to get per-word, per-head outputs:
```python
with daam.trace(model) as trc:
    output_image = model(**settings).images[0]
    heads = len(trc.all_heat_maps.heads())
    display(output_image)
    for word in prompt.split(" "):
        word = re.sub(r'[^\w]', '', word)
        fig, ax = make_im_subplots(math.ceil(heads / 4), 4)
        fig.suptitle(word)
        for head_idx in range(heads):
            ax_ = ax[head_idx // 4, head_idx % 4]
            global_heat_map = trc.compute_global_heat_map(head_idx=head_idx, layer_idx=2, normalize=True) # why layer_idx=2,? # cuz q=0, k=1, v=2, o=3 ?
            word_heat_map = global_heat_map.compute_word_heat_map(word)
            word_heat_map.plot_overlay(output_image, ax=ax_)
            ax_.set_title(f'Head {head_idx}')
        fig.show()
```

**Comment by xuyang06** [comment]
Awesome idea! Do we have any plan to implement this into some stable diffusion webui plugin, e.g., https://github.com/d8ahazard/sd_dreambooth_extension ?

**Comment by runner22k** [comment]
Great. will this help in speeding up the training process or Memory usage or Quality of the training or all?

**Comment by ima9ine** [comment]
Hi! I have a question regarding masking training,
I just checked a function loss_step of [cli_lora_pti.py](https://github.com/cloneofsimo/lora/blob/bdd51b04c49fa90a88919a19850ec3b4cf3c5ecd/lora_diffusion/cli_lora_pti.py#L132)
I just want to check that you used same loss_step with masked loss calculation not only for Tuning but also for "PIT".
I agree idea using masking for PIT since it gonna help localizing. 
Thank you for great work!

**Comment by landmann** [comment]
Hi all! Anyone can point me at what process was used to do this? I love the idea behind it - I'd love to implement it.

Also, why do we have to add the mask latents onto the model_pred input? Haha, a lot to parse through...

https://github.com/cloneofsimo/lora/discussions/271#discussion-6519912
